Add the parallel mode for the North American multi-site SDA workflow.

DongchenZ opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Add the parallel mode for the North American multi-site SDA workflow.

The workflow contains functions to split site lists into various chunks and run SDA for each group of sites.
<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [ ] Immediately
- [ ] Within one week
- [x] When possible

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] My name is in the list of CITATION.cff
- [ ] I agree that PEcAn Project may distribute my contribution under any or all of
	- the same license as the existing code,
	- and/or the BSD 3-clause license.
- [x] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

Commit:
Add parallel SDA workflow.

Commit:
Update change log.

mdietze commented on modules/assim.sequential/R/sda.enkf_parallel.R:
```diff
@@ -0,0 +1,608 @@
+    settings$modeloutdir <- file.path(outdir, "out")
+    settings$host$folder <- file.path(outdir, "out")
+    settings$host$outdir <- file.path(outdir, "out")
+    settings$host$rundir <- file.path(outdir, "run")
```
This chunk is silently making very strong assumptions that various folders are always in their default locations, which seems dangerous if they are not. Also, setting the local and host as identical won't work if the host isn't the localhost.

DongchenZ replied:
Fixed the first part by adding a couple of  `logger.info`. For the second part, the current function is called `sda.enkf_local`, which means it can only be run locally. I also added the messaging part to warn the user if they would like to run it through the remote server.

mdietze commented on modules/assim.sequential/R/sda.enkf_parallel.R:
```diff
@@ -0,0 +1,608 @@
+    jobsh <- gsub("@FOLDER_PATH@", folder.path, jobsh)
+    writeLines(jobsh, con = file.path(folder.path, "job.sh"))
+    # qsub command.
+    qsub <- "qsub -l h_rt=10:00:00 -l mem_per_core=4G -l buyin -pe omp @CORES@ -V -N @NAME@ -o @STDOUT@ -e @STDERR@ -S /bin/bash"
```
These bits are hard coded to BU settings. Anything that's BU specific needs to come from the settings file.

DongchenZ replied:
Fixed.

mdietze commented on modules/assim.sequential/R/sda.enkf_parallel.R:
```diff
@@ -0,0 +1,608 @@
+#' @return NONE
+#' @export
+#' 
+sda.enkf_local <- function(settings, 
```
At a high level, there's too much code internal to this function that is identical to that in the SDA multisite. That's going to make both too hard to maintain -- any change in one will need to be made in the other and tested. Much better approach would be to refactor so as to put the shared parts in shared functions that are called by both.

DongchenZ replied:
For this PR, I would like to keep this version for now (so I can reference the exact version of pecan in Chapter 2, and this PR will not affect other people's SDA workflow). And you are right, in the longer term, we will definitely merge those two SDA functions. However, the code-cleaning and merging are non-trivial tasks. I will open another PR that incorporates both functions with a cleaner format, comments, logics, etc.

Commit:
Resolve conflicts.

Merge branch 'develop' of https://github.com/PecanProject/pecan into SDA_data

# Conflicts:
#	CHANGELOG.md

Commit:
Update sda parallel function.

Commit:
add unlist to extract the path.

Commit:
Update phenology process.

Commit:
Update wood initialization.

Commit:
Update arguments.

Commit:
Update arguments.

Commit:
Update parallel computation in the SDA analysis block function.

Commit:
Update CPU detection.

Commit:
Update analysis block function.

Commit:
Bug fix for the dimensionality in Bayesian sampling.

Commit:
Update the script.

Commit:
Improve the printing-out control.

DongchenZ committed:
Merge branch 'PecanProject:develop' into SDA_data

Commit:
Resolve Mike's suggestions.

Commit:
Added the runner script.

Commit:
Remove cores argument.

Commit:
Update cores detection.

DongchenZ commented:
@mdietze @infotroph @dlebauer @Qianyuxuan @cami-webb 
This PR is for the continental SDA workflow, which also made some changes to the current SIPNET SDA functions. Please review the changes and let me know if you have any comments or suggestions.

Specifically, the changes are mostly due to the edge cases across the continent:
1. In `models/sipnet/R/read_restart.SIPNET.R` script.
    1) Remove the `print(run.id)` for clearner log output.

2. In `models/sipnet/R/write.configs.SIPNET.R` script.
    1) resolve the edge cases where locations may have a `leafoff` date earlier than a `leafon` date (See Line 477 - 487).
    2) resolve an issue where SIPNET takes plantWoodInit as all woods (including roots) when t=1, which will lead to smaller wood than expected (See Line 641 - 644).

3. In `models/sipnet/R/write_restart.SIPNET.R` script.
    1) Add an `if-control` for printing out the message (See Line 40 - 42).

4. In `modules/assim.sequential/R/Analysis_sda_block.R` script.
    1) Grabbing `cores` from settings to standardize the parallel computation throughout the SDA workflow (See Line 21 - 28).
    2) Replacing `furrr::map` with `foreach` to silence the MCMC messages (See Line 69 - 89).
    3) Returning the `analysis` output per block to avoid heavy computation load when we sample across the continent (See Line 106 - 107, 668 - 673, and 683).
    4) Resolving edge cases when there are limited observations (<= 1) for a specific site (See Line 243 - 246) when building blocks for the MCMC sampling (See Line 245 - 248).
    5) Resolving edge cases where the process error (in precision) could be very small, causing the singularity problem during the MCMC calculations (this was hard-coded long ago; I will test if the problem still exists, and I think this will not affect anyone) (See Line 411 - 415).
    6) Resolving the edge cases where we have limited observations (<= 1) during the calculations of `mufa` and `pfa` after the MCMC sampling (See Line 542 - 552).

5. In `modules/assim.sequential/R/Nimble_codes.R` script.
    1) Resolving edge cases where we have only one observation for a specific site (See Line 214).

6. In `modules/uncertainty/R/ensemble.R` script.
    1) Adding `unlist` to the path in case the path is in the `list` format instead of a vector (mine is in the `list` format, which causes an error when trying to read the IC file).

Commit:
Remove namespace.

Commit:
Update documentation.

infotroph commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -315,6 +344,7 @@ build.block.xy <- function(settings, block.list.all, X, obs.mean, obs.cov, t) {
   #if it's Wishart Q, we need to replace any NA Y with corresponding muf, and r with Pf.
   #also, if length of observation is 1, the Wishart Q is not suitable for the MCMC.
   #we will then need to change the Q type to 3, which is the vector Q.
+  #the wishart-MCMC is still under development, so I commanded them out for now.
```
```suggestion
  #the wishart-MCMC is still under development, so I commented them out for now.
```

DongchenZ replied:
Fixed.

infotroph commented on modules/assim.sequential/R/sda.enkf_parallel.R:
```diff
@@ -0,0 +1,644 @@
+    if (!is.null(obs.mean[[t]][[1]]) | (as.logical(settings$state.data.assimilation$free.run) & control$forceRun)) {
+      #decide if we want the block analysis function or multi-site analysis function.
+      #initialize block.list.all.
+      if (t == 1 | !exists("block.list.all")) {
```
```suggestion
    if (!is.null(obs.mean[[t]][[1]]) || (as.logical(settings$state.data.assimilation$free.run) & control$forceRun)) {
      #decide if we want the block analysis function or multi-site analysis function.
      #initialize block.list.all.
      if (t == 1 || !exists("block.list.all")) {
```

DongchenZ replied:
Fixed.

infotroph commented on modules/assim.sequential/R/sda.enkf_parallel.R:
```diff
@@ -0,0 +1,644 @@
+    if(!file.exists(file.path(f, "sda.all.forecast.analysis.Rdata"))) {
+      return(f)
+    }
+  }) %>% unlist)
```
This looks like it works fine, but FTR since file.exists accepts vectors I think you could skip the `map` entirely:

```suggestion
fail.folders <- folders[!file.exists(file.path(folders, "sda.all.forecast.analysis.Rdata"))]
```

DongchenZ replied:
Thanks for the suggestion!

infotroph commented on modules/uncertainty/R/ensemble.R:
```diff
@@ -464,7 +464,7 @@ for (input_tag in names(settings$run$inputs)) {
      if (!input_tag %in% names(samples)) {
         # Use first path (already validated as single path)
-        settings$run$inputs[[input_tag]]$path <- input$path[1]  } 
+        settings$run$inputs[[input_tag]]$path <- unlist(input$path[1])} 
```
Just to confirm: This has been tested after #3561, correct? I think they should be compatible (That was just for the one-input case whereas this will also work on each member of a larger ensemble), but wanted to check.

DongchenZ replied:
Yes. I tested using the newest version of pecan.

infotroph commented on CHANGELOG.md:
```diff
@@ -42,6 +42,8 @@ section for the next release.
 - added support for soil temperature, relative humidity, soil moisture, and PPFD downscaling to `met_temporal_downscale.Gaussian_ensemble`
 
+- Added parallel mode for the entire SDA workflow.
+
```
This is now in an already-released section -- Move up above line 15

DongchenZ replied:
Fixed.

infotroph reviewed (approved):
Looks good to me! Glad to be converging on a unified parallel strategy. I left a couple very minor comments; only the one about changelog line placement is essential.

Note that I focused on the mechanics of the parallel changes and am leaving all consideration of how it will affect the SDA results to you and @mdietze.

Commit:
Fix github check.

DongchenZ committed:
Merge branch 'PecanProject:develop' into SDA_data

Commit:
Merge branch 'SDA_data' of https://github.com/DongchenZ/pecan into SDA_data

Commit:
Resolve GitHub suggestions.

Commit:
Resolve github issue.

infotroph commented:
@mdietze Have your requested changes been addressed here?

mdietze reviewed (approved):
Approving, but would like to see an Issue opened about pulling together (and probably refactoring) all the different SDA code variants. It makes me nervous to see the number of variants keep going up instead of going down!

infotroph committed:
Merge branch 'develop' into SDA_data
